lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2508,6 +2516,116 @@ signTransaction tl preferredLatestEra keyLookup (rootKey, rootPwd) utxo =
+                    $ availableUTxO pendingTxs checkpoint
+                , checkpoint
+                , pendingTxs
+                )
+
+        protocolParams <- currentProtocolParameters netLayer
+        unsignedTxBody <-
+            either (throwIO . (ExceptionConstructTx . ErrConstructTxBody)) pure $
+                mkUnsignedTransaction txLayer @recentEra
+                    (getRawKey (Seq.rewardAccountKey (getState wallet)))
+                    protocolParams txContext (Left (PreSelection []))
+
+        unsignedBalancedTx <- balanceTx protocolParams utxo unsignedTxBody
+        signedSealedTx <- signBalancedTx wallet unsignedBalancedTx
+
+        let ( tx
+                , _tokenMapWithScripts1
+                , _tokenMapWithScripts2
+                , _certificates
+                , _validityIntervalExplicit
+                , _witnessCount
+                ) = decodeTx txLayer anyCardanoEra signedSealedTx
+        let amountIn = F.fold $ snd <$> tx ^. #resolvedInputs
+        let amountOut = F.fold $ TxOut.coin <$> tx ^. #outputs
+        let meta =
+                mkTxMeta
+                    (currentTip wallet)
+                    (txValidityInterval txContext)
+                    amountIn
+                    amountOut
+
+        pure (tx, meta, signedSealedTx)
+  where
+    anyCardanoEra = WriteTx.fromAnyRecentEra era
+
+    balanceTx
+        :: WriteTx.IsRecentEra era
+        => ProtocolParameters
+        -> (UTxOIndex WalletUTxO, Wallet (SeqState n ShelleyKey), Set Tx)
+        -> Cardano.TxBody era
+        -> IO (Cardano.Tx era)
+    balanceTx protocolParams utxo unsignedTxBody = do
+        let protocolParameters =
+                ( protocolParams
+                , fromJust $ currentNodeProtocolParameters protocolParams
+                )
+        let partialTx = PartialTx
+                (Cardano.Tx unsignedTxBody []) (Cardano.UTxO mempty) mempty
+        balanceTransaction @_ @_ @(SeqState n ShelleyKey)
+            tr txLayer (delegationAddress @n @ShelleyKey @'CredFromKeyK)
+                protocolParameters ti utxo partialTx
+                & (runExceptT >=> either (throwIO . ExceptionBalanceTx) pure)
+
+    signBalancedTx
+        :: Cardano.IsCardanoEra era
+        => Wallet (SeqState n ShelleyKey) -> Cardano.Tx era -> IO SealedTx
+    signBalancedTx wallet unsignedBalancedTx =
+        runExceptT ( do
+            let wrapError = ErrWitnessTxWithRootKey
+            withRootKey db walletId pwd wrapError $ \rootKey scheme -> do
+                let keyWithPassphrase = (rootKey, preparePassphrase scheme pwd)
+                pure $ signTransaction @_ @'CredFromKeyK txLayer
+                    anyCardanoEra
+                    (isOwned (getState wallet) keyWithPassphrase)
+                    keyWithPassphrase
+                    (wallet ^. #utxo)
+                    (sealedTxFromCardano $ inAnyCardanoEra unsignedBalancedTx)
+            )
+        >>= either (throwIO . ExceptionWitnessTx) pure
```

Comment:
We can avoid the multi-line `( do ...)` with `either (throwIO . ExceptionWitnessTx) pure <=< runExceptT $ do`:

```suggestion
    signBalancedTx wallet unsignedBalancedTx =
        either (throwIO . ExceptionWitnessTx) pure <=< runExceptT $ do
            let wrapError = ErrWitnessTxWithRootKey
            withRootKey db walletId pwd wrapError $ \rootKey scheme -> do
                let keyWithPassphrase = (rootKey, preparePassphrase scheme pwd)
                pure $ signTransaction @_ @'CredFromKeyK txLayer
                    anyCardanoEra
                    (isOwned (getState wallet) keyWithPassphrase)
                    keyWithPassphrase
                    (wallet ^. #utxo)
                    (sealedTxFromCardano $ inAnyCardanoEra unsignedBalancedTx)
```